onchain/rollups/test/foundry/consensus/quorum/QuorumFactory.t.sol

```diff
@@ -0,0 +1,121 @@
+// (c) Cartesi and individual authors (see AUTHORS)
+// SPDX-License-Identifier: Apache-2.0 (see LICENSE)
+
+/// @title Quorum Factory Test
+pragma solidity ^0.8.8;
+
+import {Test} from "forge-std/Test.sol";
+import {QuorumFactory} from "contracts/consensus/quorum/QuorumFactory.sol";
+import {Quorum} from "contracts/consensus/quorum/Quorum.sol";
+import {IHistory} from "contracts/history/IHistory.sol";
+import {Vm} from "forge-std/Vm.sol";
+
+contract QuorumFactoryTest is Test {
+    QuorumFactory factory;
+    address[] validators;
+    uint256[] shares;
+    uint256 numValidator;
+
+    event QuorumCreated(Quorum quorum, IHistory history);
+
+    function setUp() public {
+        factory = new QuorumFactory();
+    }
+
+    function testNewQuorum(uint16 _numValidator, IHistory _history) public {
+        generateAddrAndShares(_numValidator);
+
+        vm.recordLogs();
+
+        Quorum quorum = factory.newQuorum(validators, shares, _history);
+
+        testNewQuorumAux(validators, shares, _history, quorum);
+    }
+
+    function generateAddrAndShares(uint256 _numValidator) internal {
+        // assume there are less than 1000 validators so the test wont run forever
+        vm.assume(_numValidator > 0 && _numValidator < 1000);
+        numValidator = _numValidator;
+
+        validators = new address[](numValidator);
+        shares = new uint256[](numValidator);
+        for (uint256 i; i < numValidator; ++i) {
+            validators[i] = vm.addr(i + 1); // assume there is no address collision
+            shares[i] = i + 1;
+        }
+    }
+
+    function testNewQuorumAux(
+        address[] memory _validators,
+        uint256[] memory _shares,
+        IHistory _history,
+        Quorum _quorum
+    ) internal {
+        Vm.Log[] memory entries = vm.getRecordedLogs();
+
+        uint256 numQuorumCreated;
+        for (uint256 i; i < entries.length; ++i) {
+            Vm.Log memory entry = entries[i];
+
+            if (
+                entry.emitter == address(factory) &&
+                entry.topics[0] == QuorumCreated.selector
+            ) {
+                ++numQuorumCreated;
+                (address quorumAddr, address historyAddr) = abi.decode(
+                    entry.data,
+                    (address, address)
+                );
+                assertEq(quorumAddr, address(_quorum));
+                assertEq(historyAddr, address(_history));
+            }
+        }
+        assertEq(numQuorumCreated, 1);
+
+        assertEq(_quorum.numOfValidators(), numValidator);
+        assertEq(address(_quorum.getHistory()), address(_history));
+        address[] memory returnedValidators = _quorum.validators();
```

Comment:
We should check here that:
- the array of validators returned by the Quorum contract (via `validators`) should have the expected length (`numValidator`)
- the total number of shares returned by the Quorum contract (via `totalShares()`) should be the expected value (sum of `_shares`)